otherlibs/unix/spawn.c

```diff
@@ -0,0 +1,138 @@
+
+CAMLprim value unix_spawn(value executable,
+                          value args,
+                          value optenv,
+                          value usepath,
+                          value redirect)
+{
+  char ** argv;
+  char ** envp;
+  const char * path;
+  pid_t pid;
+  int src, dst, r;
+  posix_spawn_file_actions_t act;
+
+  caml_unix_check_path(executable, "create_process");
+  path = String_val(executable);
+  argv = cstringvect(args, "create_process");
+  if (Is_block(optenv)) {
+    envp = cstringvect(Field(optenv, 0), "create_process");
+  } else {
+    envp = environ;
+  }
+  posix_spawn_file_actions_init(&act);
+  for (dst = 0; dst <= 2; dst++) {
+    src = Int_val(Field(redirect, dst));
+    if (src != dst) {
+      r = posix_spawn_file_actions_adddup2(&act, src, dst);
+      if (r != 0) goto error;
+      r = posix_spawn_file_actions_addclose(&act, src);
+      if (r != 0) goto error;
+    }
+  }
+  if (Bool_val(usepath)) {
+    r = posix_spawnp(&pid, path, &act, NULL, argv, envp);
+  } else {
+    r = posix_spawn(&pid, path, &act, NULL, argv, envp);
+  }
+ error:
+  posix_spawn_file_actions_destroy(&act);
+  cstringvect_free(argv);
+  if (Is_block(optenv)) cstringvect_free(envp);
+  if (r != 0) unix_error(r, "create_process", executable);
+  return Val_long(pid);
+}
+
+#else
+
+#ifndef HAS_EXECVPE
+extern int unix_execvpe_emulation(const char * name,
+                                  char * const argv[],
+                                  char * const envp[]);
+#endif
+
+#define ERROR_EXIT_STATUS 127
+
+CAMLprim value unix_spawn(value executable,
+                          value args,
+                          value optenv,
+                          value usepath,
+                          value redirect)
+{
+  char ** argv;
+  char ** envp;
+  const char * path;
+  pid_t pid;
+  int src, dst;
+
+  caml_unix_check_path(executable, "create_process");
+  path = String_val(executable);
+  argv = cstringvect(args, "create_process");
+  if (Is_block(optenv)) {
+    envp = cstringvect(Field(optenv, 0), "create_process");
+  } else {
+    envp = NULL;
+  }
+  pid = fork();
+  if (pid == -1) {
+    cstringvect_free(argv);
+    if (envp != NULL) cstringvect_free(envp);
+    uerror("create_process", executable);
+  }
+  if (pid == 0) {
+    for (dst = 0; dst <= 2; dst++) {
+      src = Int_val(Field(redirect, dst));
+      if (src != dst) {
+        if (dup2(src, dst) == -1) _exit(ERROR_EXIT_STATUS);
+        if (close(src) == -1) _exit(ERROR_EXIT_STATUS);
+      }
+    }
+    if (Bool_val(usepath))
+      if (envp == NULL)
+        execvp(path, argv);
+      else
+#ifdef HAS_EXECVPE
+        execvpe(path, argv, envp);
+#else
+        unix_execvpe_emulation(path, argv, envp);
+#endif
+    else
+      if (envp == NULL)
+        execv(path, argv);
+      else
+        execve(path, argv, envp);
+    _exit(ERROR_EXIT_STATUS);
+  }
+  cstringvect_free(argv);
+  if (envp != NULL) cstringvect_free(envp);
+  return Val_long(pid);
```

Comment:
Nitpick: I think the code would be easier to read if this parent cleanup&return logic was a `if (pid != 0)` right after the `if (pid == -1)` case, so that the early-exit cleanup logics are close together, with the exec-in-child case being the main body of the function. (One could even have a single case `if (pid != 0)` and then error on `-1` before returning, as you do in the posix_spawn version, but that could also be confusing.)

Reply:
Fair enough.  I tried the alternate style in c2ab1c331 .